### ISSUE TITLE
DataLoader indices[] under-filled: epoch sees ~3% of dataset

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#91]
**Audit context:** Filed as part of a USERAPI audit of ODT (Pass 1, April 2026).
Line numbers below reference `es-ude/OnDeviceTraining@b0a737a9d1ff71d71187c6600b56ed5eff4eaab0`.

## Summary

Two layers of the DataLoader disagree about the size of the `indices[]` array. The outer layer allocates `sizeDataset` entries (`calloc`-zeroed); the inner layer fills only `sizeDataset / batchSize` entries. At MNIST BS=32 on 60000 samples that is 1875 filled of 60000. `getBatch` reads indices `0..sizeDataset-1`, so every sample index ≥ 1875 silently resolves to sample 0. Per-epoch unique samples: ~3% of the dataset.

## Severity

| Axis | Value |
|---|---|
| Priority | High |
| Affected users | All users whose epoch covers more than `sizeDataset / batchSize` batches (i.e. anyone doing a full-epoch pass) |
| Failure mode | Silent |
| Blast radius | Training correctness |

## Root cause

**Outer-layer allocation** — `src/src/userApi/data_loader/DataLoaderApi.c:28-29` — https://github.com/es-ude/OnDeviceTraining/blob/b0a737a9d1ff71d71187c6600b56ed5eff4eaab0/src/src/userApi/data_loader/DataLoaderApi.c#L28-L29

```c
size_t numberOfIndices = getDatasetSize();            // e.g. 60000 for MNIST train
indices = calloc(numberOfIndices, sizeof(size_t));    // zero-initialized
```

**Inner-layer fill** — `src/src/data_loader/DataLoader.c:36-40` — https://github.com/es-ude/OnDeviceTraining/blob/b0a737a9d1ff71d71187c6600b56ed5eff4eaab0/src/src/data_loader/DataLoader.c#L36-L40

```c
size_t numberOfIndices = sizeDataset / batchSize;     // e.g. 1875 at BS=32
for (size_t i = 0; i < numberOfIndices; i++) {
    indices[i] = i;
}
```

**Read-back** — `DataLoaderApi.c:54-57` — `getSampleByIndex` reads `indices[index]` for `index ∈ 0..sizeDataset-1` via `getBatch` (`sampleIndex + i` where `sampleIndex = batch_idx × batchSize`). Entries `1875..59999` are still `calloc`-zero, so they all map back to sample 0.

Per-epoch unique samples: ~1875 of 60000 = ~3% of the dataset.

## Reproducer

Source inspection is self-evident from the two pointer-to-size expressions. For empirical corroboration, instrument `getSampleByIndex` to log `indices[index]` for one full-epoch pass at BS=32 on MNIST train:

- First ~58 batches (reading `indices[0..1874]`): monotonically increasing unique values `0, 1, 2, ..., 1874`.
- All subsequent batches (reading `indices[1875..59999]`): every value is `0`.

Shuffle-invariance as corroboration: in the audit, 10-epoch MNIST training at MLP `depth=1` BS=32 produced:

| mode | PyTorch (N=5 seeds) | ODT (N=1, seed=42) | gap (ODT − PT) |
|---|---|---|---|
| shuffle | 89.69% ± 0.16 | 87.33% | −2.36 pp |
| no-shuffle | 89.67% ± 0.12 | 86.83% | −2.84 pp |

Shuffle made almost no difference (Δ = 0.50pp on ODT, 0.02pp on PyTorch) — expected if shuffle only permutes 1875 effectively-usable indices rather than 60000.

## Impact

- **Every training run in every example** using a shuffled DataLoader silently trains on ~3% of the intended dataset.
- Cross-framework accuracy comparisons are apples-to-oranges (PyTorch sees 60000 samples/epoch; ODT effectively 1875) until this is fixed.
- MNIST survives (~87% on 2-layer MLP from 1875 samples); harder datasets won't. Accuracy ceilings are silently lower than expected.
- Bug is silent because: (1) no crash — sample 0 is valid data; (2) gradient on repeated sample-0 accesses decays to near-zero once the model memorizes it, so it degenerates to a no-op rather than introducing visible instability; (3) `PRINT_ERROR` is silently compiled out at default configuration (see the separate issue on `PRINT_ERROR` wiring in this audit batch).
- Partial explanation for why shuffle-semantics appear to not matter in ODT: shuffle only reorders the 1875 usable indices per epoch.

## Recommended fix

**One-line:** change `DataLoader.c:37` from

```c
size_t numberOfIndices = sizeDataset / batchSize;
```

to

```c
size_t numberOfIndices = sizeDataset;
```

Alternative reading: if the intent at `DataLoader.c:37` was "drop last partial batch", the correct form would be `(sizeDataset / batchSize) * batchSize` — but because the loop at lines 38-40 treats `i` as a sample index (`indices[i] = i`), the 1:1 fix is correct.

## Verification

Regression test: after one epoch pass over the full dataset (e.g. MNIST train at 60000 BS=32), collect the multiset of requested sample indices via `getSampleByIndex`. Assert that unique coverage is at least 90% of the dataset size:

```c
bool seen[datasetSize] = {false};
size_t numBatches = datasetSize / batchSize;
for (size_t b = 0; b < numBatches; b++) {
    for (size_t i = 0; i < batchSize; i++) {
        size_t idx = getSampleByIndex(b * batchSize + i);
        seen[idx] = true;
    }
}
size_t unique = 0;
for (size_t i = 0; i < datasetSize; i++) if (seen[i]) unique++;
assert(unique >= datasetSize * 9 / 10);
```

Before fix: `unique ≈ 1875` (test fails). After fix: `unique == datasetSize` (test passes).

## Related

No existing upstream issues are directly related. Cross-references for this audit batch live in the audit tracking index issue.

**Comment by LeoBuron** [comment]
Duplicate of #74 (closed). The bug was fixed by commit `adc63f7 fix(#74): initDataLoader initializes all dataset indices, not just numberOfBatches` which shipped in PR #75.

Current state:
- Code: `src/data_loader/DataLoader.c:38-39` initializes the full `indices[0..sizeDataset)` range, no longer `sizeDataset / batchSize`.
- Regression test: `testGetBatch_ShuffledMinibatchCoversAllSamples` in `test/unit/data_loader/UnitTestDataLoader.c:197` collects all samples across all batches with shuffle=true and asserts full dataset coverage. This catches exactly the failure mode described in this issue.

The audit that filed this issue referenced commit `b0a737a9`, which predates PR #75. Closing as duplicate. No further action needed.
